File: next-permutation/Solution31.java
// https://leetcode.com/problems/next-permutation/

import java.util.Arrays;

class Solution31 {

    // T: O(n), S: O(1)
    public void nextPermutation(int[] nums) {
        int i = nums.length - 2;
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        if (i >= 0) {
            int j = nums.length - 1;
            while (nums[j] <= nums[i]) {
                j--;
            }
            swap(nums, i, j);
        }

        reverse(nums, i + 1);
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    private void reverse(int[] nums, int start) {
        int end = nums.length - 1;
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }

    public static void main(String[] args) {
        var sol = new Solution31();
        var nums1 = new int[] { 1, 2, 3 };
        var nums2 = new int[] { 3, 2, 1 };
        var nums3 = new int[] { 1, 1, 5 };
        sol.nextPermutation(nums1);
        sol.nextPermutation(nums2);
        sol.nextPermutation(nums3);
        System.out.println(Arrays.toString(nums1));
        System.out.println(Arrays.toString(nums2));
        System.out.println(Arrays.toString(nums3));
    }

}

File: next-permutation/solution31.go
// https://leetcode.com/problems/next-permutation/

package main

import (
	"fmt"
)

// T: O(n), S: O(1)
func nextPermutation(nums []int) {
	i := len(nums) - 2
	for i >= 0 && nums[i] >= nums[i+1] {
		i--
	}

	if i >= 0 {
		j := len(nums) - 1
		for nums[j] <= nums[i] {
			j--
		}
		nums[i], nums[j] = nums[j], nums[i]
	}

	reverse(nums, i+1)
}

func reverse(nums []int, start int) {
	end := len(nums) - 1
	for start < end {
		nums[start], nums[end] = nums[end], nums[start]
		start++
		end--
	}
}

func main() {
	nums1 := []int{1, 2, 3}
	nums2 := []int{3, 2, 1}
	nums3 := []int{1, 1, 5}
	nextPermutation(nums1)
	nextPermutation(nums2)
	nextPermutation(nums3)
	fmt.Println(nums1)
	fmt.Println(nums2)
	fmt.Println(nums3)
}
